browser/brave_cosmetic_resources_tab_helper.cc

```diff
@@ -0,0 +1,461 @@
+        std::string rule = hide_selectors_list->GetList()[i].GetString() +
+            "{display:none !important;}";
+        cosmeticFilterConsiderNewSelectors_script +=
+            "if (!window.allSelectorsToRules.has(`" +
+              hide_selectors_list->GetList()[i].GetString() + "`)) {"
+              "window.cosmeticStyleSheet.insertRule(`" + rule +
+                "`, nextIndex);"
+              "window.allSelectorsToRules.set(`" +
+                hide_selectors_list->GetList()[i].GetString() +
+                "`, nextIndex);"
+              "nextIndex++;"
+              "window.firstRunQueue.add(`" +
+                hide_selectors_list->GetList()[i].GetString() + "`);"
+            "}";
+      }
+      cosmeticFilterConsiderNewSelectors_script +=
+          "if (!document.adoptedStyleSheets.includes("
+                "window.cosmeticStyleSheet)){"
+              "document.adoptedStyleSheets = [window.cosmeticStyleSheet];"
+          "};";
+      cosmeticFilterConsiderNewSelectors_script += "})();";
+      if (hide_selectors_list->GetSize() != 0) {
+        auto* frame_host = content::RenderFrameHost::FromID(frame_id);
+        if (!frame_host)
+          return;
+        frame_host->ExecuteJavaScriptInIsolatedWorld(
+            base::UTF8ToUTF16(cosmeticFilterConsiderNewSelectors_script),
+            base::NullCallback(), ISOLATED_WORLD_ID_CHROME_INTERNAL);
+      }
+    }
+  }
+
+  std::string styled_stylesheet = "";
+  if (force_hide_selectors_list!= nullptr &&
+      force_hide_selectors_list->GetSize() != 0) {
+    for (size_t i = 0; i < force_hide_selectors_list->GetSize(); i++) {
+      if (i != 0) {
+        styled_stylesheet += ",";
+      }
+      styled_stylesheet += force_hide_selectors_list->GetList()[i].GetString();
+    }
+    styled_stylesheet += "{display:none!important;}\n";
+  }
+  base::DictionaryValue* style_selectors_dictionary = nullptr;
+  if (resources_dict->GetDictionary("style_selectors",
+      &style_selectors_dictionary)) {
+    for (const auto& it : style_selectors_dictionary->DictItems()) {
+      base::ListValue* style_selectors_list = nullptr;
+      if (!style_selectors_dictionary->GetList(it.first,
+          &style_selectors_list) || style_selectors_list->GetSize() == 0) {
+        continue;
+      }
+      styled_stylesheet += it.first + "{";
+      for (size_t i = 0; i < style_selectors_list->GetSize(); i++) {
+        if (i != 0) {
+          styled_stylesheet += ";";
+        }
+        styled_stylesheet += style_selectors_list->GetList()[i].GetString();
+      }
+      styled_stylesheet += ";}\n";
+    }
+    if (!styled_stylesheet.empty()) {
+      std::string cosmeticFilterConsiderNewSelectors_script =
+          "(function() {"
+            "let nextIndex = window.cosmeticStyleSheet.rules.length;";
+      cosmeticFilterConsiderNewSelectors_script +=
+            "window.cosmeticStyleSheet.insertRule(`" + styled_stylesheet +
```

Comment:
not 100% sure, but I don't think `insertRule` will work with a full stylesheet

Reply:
do you have any particular website example that hits that code? IIRC I had hard time hitting it.

Reply:
An easy way would be to enter multiple `:style(...)` rules in the "Custom filters" box on brave://adblock.

So for example, you could add these two lines:
```
example.com##body:style(background-color: red !important)
example.com##h1:style(color: green)
```

And then verify that https://example.com looks like this:
![2020-12-16-155236_1037x414_scrot](https://user-images.githubusercontent.com/22821309/102405182-be0c3180-3fb6-11eb-80cc-d24f0dd0c3ec.png)

Reply:
done, made insertion per rule